// atom-languageclient/build/lib/adapters/code-action-adapter.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const assert = require("assert");
const convert_1 = require("../convert");
class CodeActionAdapter {
    // Returns a {Boolean} indicating this adapter can adapt the server based on the
    // given serverCapabilities.
    static canAdapt(serverCapabilities) {
        return serverCapabilities.codeActionProvider === true;
    }
    // Public: Retrieves code actions for a given editor, range, and context (diagnostics).
    // Throws an error if codeActionProvider is not a registered capability.
    //
    // * `connection` A {LanguageClientConnection} to the language server that provides highlights.
    // * `serverCapabilities` The {ServerCapabilities} of the language server that will be used.
    // * `editor` The Atom {TextEditor} containing the diagnostics.
    // * `range` The Atom {Range} to fetch code actions for.
    // * `diagnostics` An {Array<atomIde$Diagnostic>} to fetch code actions for.
    //                 This is typically a list of diagnostics intersecting `range`.
    //
    // Returns a {Promise} of an {Array} of {atomIde$CodeAction}s to display.
    static getCodeActions(connection, serverCapabilities, linterAdapter, editor, range, diagnostics) {
        return __awaiter(this, void 0, void 0, function* () {
            if (linterAdapter == null) {
                return [];
            }
            assert(serverCapabilities.codeActionProvider, 'Must have the textDocument/codeAction capability');
            const commands = yield connection.codeAction({
                textDocument: convert_1.default.editorToTextDocumentIdentifier(editor),
                range: convert_1.default.atomRangeToLSRange(range),
                context: {
                    diagnostics: diagnostics.map((diagnostic) => {
                        // Retrieve the stored diagnostic code if it exists.
                        // Until the Linter API provides a place to store the code,
                        // there's no real way for the code actions API to give it back to us.
                        const converted = convert_1.default.atomIdeDiagnosticToLSDiagnostic(diagnostic);
                        if (diagnostic.range != null && diagnostic.text != null) {
                            const code = linterAdapter.getDiagnosticCode(editor, diagnostic.range, diagnostic.text);
                            if (code != null) {
                                converted.code = code;
                            }
                        }
                        return converted;
                    }),
                },
            });
            return commands.map((command) => ({
                apply() {
                    return __awaiter(this, void 0, void 0, function* () {
                        yield connection.executeCommand({
                            command: command.command,
                            arguments: command.arguments,
                        });
                    });
                },
                getTitle() {
                    return Promise.resolve(command.title);
                },
                // tslint:disable-next-line:no-empty
                dispose() { },
            }));
        });
    }
}
exports.default = CodeActionAdapter;
